File: lab4/lab4.jl
# Moorits Mihkel Muru, University of Tartu 2022
# Lab 4 of the Statistical Data Science (LTMS.00.025) course
# Solutions in Julia

using DataFrames      # Tables for data
using StatsBase       # Basic statistics tools
using FreqTables      # Frequency tables
using Distributions   # Normal distribution
using HypothesisTests # Hypothesis Testing
using StatsPlots      # Plotting
using HTTP            # Stream data from web
using FileIO          # Reading in a file from a web stream
using CodecZlib       # Unpacking RData file
using RData           # Reading RData files
using CSV             # Reading CSV files
using Dates           # Standard library for working with datetime formats

# Reading the data from web
students_dataset_url = "http://www.ms.ut.ee/mart/andmeteadus/students.RData"
request = HTTP.get(students_dataset_url)
buffer = CodecZlib.GzipDecompressorStream(IOBuffer(request.body))
students = DataFrame(RData.load(Stream{format"RData"}(buffer))["students"])

# T-test to infer if the height of the students is the same as in 1933.
OneSampleTTest(students.height[students.gender.==2], 173.83)

# DataFrame with no missing values in weights
weight_students = dropmissing(students, :weight) 
# Can male students' mean weight be X kg?
OneSampleTTest(weight_students.weight[weight_students.gender.==2], 75) # p-value: 0.1217 => retain H₀
OneSampleTTest(weight_students.weight[weight_students.gender.==2], 76) # p-value: 0.6011 => retain H₀

# Is the die fair?
rolls = [1, 5, 6, 6, 3, 5, 6, 6, 4, 5, 3, 6]
OneSampleTTest(rolls, 3.5) # Does not seem likely to be a fair die
confint(OneSampleTTest(rolls, 3.5), level=0.99)
pvalue(OneSampleTTest(rolls, 3.5))
# However Chi² seems to be less apt to reject the null hypothesis
ChisqTest([count(==(x), rolls) for x in 1:6])

# Chi² test for immunity to pesticides in crossed rice
ChisqTest([585, 1191, 590], [0.25, 0.5, 0.25])

# Are earthquakes caused by sinful activities?
earthquake_data_url = "http://www.ms.ut.ee/mart/andmeteadus/quake.csv"
earthquakes = DataFrame(CSV.File(IOBuffer(HTTP.get(earthquake_data_url).body)))
# Create a column with weekdays with function composing operator ∘
# First applies chop, then DateTime, then dayofweek
# Period between the parenthesis indicates broadcasting
earthquakes.weekday = (dayofweek ∘ DateTime ∘ chop).(earthquakes.time)
freqtable(earthquakes.weekday)
ChisqTest([count(==(x), earthquakes.weekday) for x in 1:7]) # It turns out, earthquakes are the sign of God's wrath
confint(ChisqTest([count(==(x), earthquakes.weekday) for x in 1:7]), level=0.995)

# Create dataframe with rows that have no missing values for weight nor medical_care
weight_medicalcare_students = dropmissing(students, [:weight, :medical_care])
# Test if there is difference in weight if the students has used medical care or not
UnequalVarianceTTest(weight_medicalcare_students.weight[weight_medicalcare_students.medical_care.==1],
    weight_medicalcare_students.weight[weight_medicalcare_students.medical_care.==0])

# Are the earthquakes on Tuesdays as strong as on other days?
UnequalVarianceTTest(earthquakes.mag[earthquakes.weekday.==2], earthquakes.mag[earthquakes.weekday.!=2])
# p-value 0.0803 says that they are a bit weaker, but not with a 95% confidence level

# Do students, who have needed the help of an ambulance in the past have smaller probability of rating their health as very good?
freqtable(students.ambulance, students.health .== "very good")
# I don't know what the equivalent of prop.test is...


# Chi² test for comparing study variable distribution
prop(freqtable(students, :gender, :beer), margins=1).*100

File: lab4/lab4.R
# Statistical Data Science
# 4. lab session

# Hypothesis testing -------

## Hypothesis on population means -------

# Load the data
load(url("http://www.ms.ut.ee/mart/andmeteadus/students.RData"))
attach(students)

# Simple one sample hypothesis testing of mean
t.test( height[gender==2] , mu=173.83)
t_stat = (mean(height[gender==2])-173.83)/sd(height[gender==2])*sqrt(length((height[gender==2])))

# Exercise 1 -----
# Answer the following questions:
## A. Can male students’ mean weight be 75kg?


## B. Can male students’ mean weight be 76kg?


## C. Are the results contradictive? Why?


# Exercise 2 -----

# Is it possible that John has a fair die?
die_result = c()


# Additionally, to p-value you should also think 
# of other aspects of the exercise. What can they be?



## Hypothesis on probabilities ----

# Chi-square test
chisq.test(c(585, 1191, 590), p=c(0.25, 0.5, 0.25))

# Exercise 3
data=read.csv(url("http://www.ms.ut.ee/mart/andmeteadus/quake.csv"),
              header=TRUE)


data$date=as.Date(data$time)
data$weekday=weekdays(data$date)
data[1:3,]
table(data$weekday)

# If you not used to Estonian yet, then you might want to 
# find weekdays using the following command
data$weekday=factor(c("Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
)[as.POSIXlt(data$date)$wday + 1], 
levels = c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"))

# If you want the weekdays to appear in order (in frequency tables or plots)
data$weekday=factor(data$weekday,
                    levels = c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday")
)

# Chi-square test
# Frequency table of earthquakes by weekdays
freqs = table(data$weekday)
freqs

# Probabilities under null-hypothesis - 
# we would expect to see the same number of earth quakes on every day,
# therefore probability of an earth quake on a single day would be 1/7
# and to vectorise it we can usethe rep() command
expected_prob = #your code
  expected_prob


# Now the chi-square test
chisq.test(freqs, p=expected_prob) 


# Testing two populations ----

## Comparing means----

# Mean weights of students who have had hospitalisation in the past 
# (medical_care=1) and the ones who have not (medical_care=0)
t.test(weight[medical_care==1], weight[medical_care==0])

# Exercise 4 ----
# your code

# Comparing probabilities ----
table(ambulance, health=="very good")
prop.test(c(52, 13), c(52+355, 13+60))


# Chi-square test for comparing study variable distribution
prop.table(table(gender, beer),1)*100
chisq.test(table(gender, beer))

# Expected frequencies
chisq.test(table(gender, beer))$expected

# Classify all students who drink more than 5 bottles a week 
# into a single (“drunkards”) group
beer4=factor(beer, labels=c("0", "0..1", "1..4", "5..", "5.."))
table(gender, beer4)
chisq.test(table(gender, beer4))
plot(table(beer4, gender), col=c("pink", "skyblue"))

# more formatted plot
students$genderF=factor(gender, levels=1:2,
                        labels=c("female", "male"))
windows(width=6, height=4)
barplot(prop.table(table(genderF, beer),2)*100,
        col=c("pink", "skyblue"), ylab="%", legend=T,
        xlim=c(0,8), xlab="Bottles drunk in a week",
        main="Relation between gender and beer consumption")


# Exercise 5
# your code

